### PR TITLE
Make contiguous case for `chainerx::AddAt` faster

### PR DESCRIPTION
**Pull request by emcastillo**
Reworked it similar to #8295 

IndexableArrays uses the 1D specialized template whenever it is possible.

For 40 runs of a pathological case, execution time decreased from 0.18 seconds to 0.02

Please note that `addat` is not currently exposed to the python API.

```python
def bench(name, ary_fn):
    np.random.seed(42)
    x = np.zeros((20, 96, 132)).astype(np.float32)
    y = np.random.rand(20, 96, 36, 11).astype(np.float32)
    indices = np.random.randint(132, size=(36, 11)).astype(np.int32)
    a = chx.array(x, device='cuda:0')
    b = chx.array(y, device='cuda:0')
    indices = chx.array(indices, device='cuda:0')

    for i in range(10):
        cuda = chx.addat(a,indices,b,axis=2)

    cupy.cuda.device.Device().synchronize()
    start = time.time()
    for i in range(40):
        cuda = chx.addat(a,indices,b,axis=2)
    cupy.cuda.device.Device().synchronize()
    take = time.time() - start
    print(name, take)
```

### CODE REVIEW

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 1]
```diff
@@ -139,7 +139,7 @@ __global__ void AddAtCudaKernel(
         CHAINERX_ASSERT(0 <= index);
         CHAINERX_ASSERT(index < target_dim);
         if (index == k) {
-            int bi = (i * b_right_dim * b_dim + it_indices.raw_index()) * right_dim + l;
+            int bi = (i * b_right_dim * b_dim) + it_indices.raw_index() * right_dim + l;
```
Comment (emcastillo) [inline chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: TODO: Parenthesis not needed

**Comment by emcastillo** [comment]
PTAL 😄

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 2]
```diff
@@ -81,78 +81,52 @@ __global__ void TakeCudaKernel(
...
+        IndexableArray<const TIndex, kNdim> indices,
+        Indexer<kNdim> b_indexer,
+        Indexer<kNdim> out_indexer,
+        Indexer<kNdim> indices_indexer,
+        TIndex target_dim,
+        TIndex b_dim,
+        TIndex right_dim,
+        TIndex b_right_dim,
+        TIndex num_iters,
+        IndexBoundsMode mode) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: To avoid the branch in L112, how about passing by template parameter?

**Comment by asi1024** [comment]
Jenkins and flexCI, test this please.

**Comment by asi1024** [comment]
Jenkins and flexCI, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**